### PR TITLE
WIP: Generalized N-dimensional dual number implementation

### PR DESCRIPTION
**Pull request by jrevels**
This PR implements the `DualN{N,T}` type, which stores epsilon components in an `NTuple{N,T}`. I basically ported and naively generalized the code for `Dual4`, so there's still a bit left to be done:

- [ ] Figure out the right way to handle mixed dimensional cases (e.g. `f{A,B}(::DualN{A}, ::Dual{B})` where `A != B`)
- [ ] Pretty printing for `DualN`
- [ ] Replace the currently used `map`s  with inline code for the sake of performance (I'm assuming this performance difference hasn't been fixed as of v0.4).
- [ ] Utilize `@simd` and `@inbounds` where safe/relevant in the inlined code from the previous checkbox
- [ ] Figure out some better/more tests - I basically ripped the current ones from `Dual4`'s tests.
- [ ] Figure out some solid benchmarks and test against `Dual4`

This is also my first time really working in v0.4 besides a few compatibility testing situations, so apologies if my code is "v0.3ish". Let me know what you think!

Edit: Added a couple more checkboxes. Let me know which of the above goals (if any) should be split up into a different PR.

### CODE REVIEW

File: src/dual_n.jl [hunk 1]
```diff
@@ -1,174 +1,193 @@
...
+eps(z::DualN) = eps(real(z))
+eps{N,T}(::Type{DualN{N,T}}) = eps(T)
+
+# Univariate functions that map
+# onto the DualN's real component
+# and return a DualN
+for f in (:eps, :one, :zero)
+    @eval begin
+        ($f)(z::DualN) = DualN(($f)(real(z)))
+        ($f){N,T}(::Type{DualN{N,T}}) = DualN(($f)(T))
```
Comment (mlubin) [inline src/dual_n.jl]: Does ``zero(DualN{3,Float64})`` return a ``DualN{3,Float64)`` or a ``DualN{0,Float64}`` here?

File: src/dual_n.jl [hunk 2]
```diff
@@ -1,174 +1,193 @@
...
-convert{T<:Real}(::Type{Dual4{T}}, z::Dual4{T}) = z
-convert{T<:Real}(::Type{Dual4{T}}, z::Dual4) =
-  Dual4{T}(real(z), epsilon1(z), epsilon2(z), epsilon3(z), epsilon4(z))
+isreal(z::DualN) = any(x -> x == 0, epsilon(z))
+isreal(z::DualN{0}) = true
+isfinite(z::DualN) = isfinite(real(z))
 
-convert{T<:Real}(::Type{T}, z::Dual4) =
-  (epsilon1(z)==0 && epsilon2(z) == 0 && epsilon3(z) == 0 && epsilon4(z) == 0 ? convert(T, real(z)) : throw(InexactError()))
+# Define conversion/promotion for mixed N? Is there a
```
Comment (mlubin) [inline src/dual_n.jl]: I don't think there's any need to worry about this for now

**Comment by mlubin** [comment]
This is definitely a good start. As I commented, I wouldn't worry about interactions with dual numbers with differing numbers of components. It's okay to leave this undefined for now because I don't immediately see how it would be useful. Converting the ``map`` code into efficient inline code will probably be the trickiest part.

Depending on how this goes, it's not clear that we will eventually decide to keep this code in the DualNumbers package. I would go ahead and start a fresh package for development purposes, and at the end we can decide if the code should go into DualNumbers, ForwardDiff, or somewhere else.

**Comment by scidom** [comment]
I agree @mlubin, with both of your remarks (on duals with differing numbers of components and on having a dev repo). It would be good if we set this up so that we can easily clone the commits of @jrevels and play around with the code whenever needed (it is less direct to do this from an unmerged PR).

File: src/dual_n.jl (hunk 1 above)
Comment (jrevels) [inline src/dual_n.jl]: Oh, good catch! Just pushed a fix.

**Comment by jrevels** [comment]
Sounds like a good idea! Should I host this package as a repo under my own account, or are you guys going to make one for me under JuliaDiff? Also, what should we call it?

**Comment by scidom** [comment]
How about `TmpNDual`, `NDualDev`, `NDualScratch`, `NDualWorkspace` or `NDualJSoC15`? These are some candidate names that hopefully sound tidy. It doesn't really matter, you may choose any name you like @jrevels since this is going to be a dev repo till we co-decide where your contribution will finally live.

@mlubin, @jrevels I would suggest (and can do this if you both agree) to set up a new package under JuliaDiff, with all three of us having admin access. Let me know what you think.

P.S. @jrevels on a complete separate note, I was looking at your github profile and found out that you have some very interesting packages for quantum computing in Julia. Have you ever worked on Quantum Monte Carlo (QMC) methods?

**Comment by mlubin** [comment]
I'd just call it ``DualN`` for now. JuliaDiff would be a good space since we're hosting the project.

CC @davidanthoff

**Comment by mlubin** [comment]
Done: https://github.com/JuliaDiff/DualN.jl

**Comment by scidom** [comment]
It is not a big deal, just a thought only, would it feel more Julian/natural to call the package (and associated type) `NDual`? I am thinking along the lines of tuples for example, Julia has `Tuple` and `NTuple` for instance.

**Comment by mlubin** [comment]
That's a fair point, though "n-tuple" is pretty common in mathematical vocabulary, while "n-dual" isn't, so the situation isn't quite analogous.

**Comment by scidom** [comment]
It is true that n-tuple is a proverbial mathematical term - I was thinking mostly along the verbose lines of n-dual being the most natural shorthand for n-dimensional dual, so (from a literate rather than strictly mathematical point of view) `NDual` seems more naturel (to me!). It's ok if we leave it as `DualN`, mostly expressed a preference.

**Comment by mlubin** [comment]
I don't have a strong preference here either, happy to revisit the issue before the code becomes final.

**Comment by jrevels** [comment]
Also, I'm assuming the name of the package module for will match the name of the repo, such that the name of the eponymous type would cause a conflict (e.g. if the module is called `DualN`, defining a `DualN` type would be an invalid redefinition). I'm not tied to any particular module/type name, just throwing that out there. So with the current repo name, what should I call the type/module?

@scidom I haven't, besides doing a few cursory exercises by hand to be barely familiar with the subject. Eventually I'll probably dig deeper, though, as  [JuliaQuantum](https://github.com/JuliaQuantum) should , at some point, develop functionality related to QMC methods. That might eventually be covered under the [QuDynamics](https://github.com/JuliaQuantum/QuDynamics.jl) package, or even be it's own package.

**Comment by mlubin** [comment]
@jrevels ah, right. What about renaming the module to ``NDuals.jl``? That sounds better than ``DualsN.jl``.

**Comment by scidom** [comment]
@jrevels, good to hear, just asked because of having potential collaborators in my environment with whom I may work on QMC in the near future, in which case we could co-code (I am into MCMC, though haven't done any work on QMC so far).

@mlubin I am ok with any name we choose in the end.

**Comment by jrevels** [comment]
Awesome! I'll start setting up the package today then. 

Quick workflow question: Would it be preferable for me to always group my work as PRs to the NDuals repo or should I just push directly and then you guys can review at any time (I have no preference)?

**Comment by mlubin** [comment]
I think it makes sense to push the basic functionality and do PRs for the major changes, but whatever works for your development flow. The code should be short enough that reviewing it just by reading through the source shouldn't be difficult.

**Comment by jrevels** [comment]
Sounds good.

@scidom I'm always looking for projects to learn more quantum theory! If the opportunity to collaborate arises, please do let me know.

**Comment by jrevels** [comment]
Going to go ahead and close this now that the code can be found in the [NDuals](https://github.com/JuliaDiff/NDuals.jl) package.

**Comment by davidanthoff** [comment]
Great to see this happening!

**Comment by scidom** [comment]
@jrevels I will certainly let you know once I get to work on QMC. This is streamlined to happen after a while, probably in 9 months's time from now :)

@mlubin thanks for changing the package name. In fact, adding this extra "s" at the end makes `NDuals` even more "correct" conceptually as a package name.

**Comment by scidom** [comment]
P.S. Aat the same time it allows usage of `NDual` for the actual type as @jrevels spotted, which is great!

**Pull request closed without merging**